### PR TITLE
Add support for GPUArrays 8

### PR DESCRIPTION
**Pull request by vchuravy**

### CODE REVIEW

File: src/device/runtime.jl [hunk 1]
```diff
@@ -84,10 +84,14 @@ function load_and_link!(mod, path)
...
-    #     delete!(attrs, StringAttribute("target-features"; ctx))
-    # end
+    for f in LLVM.functions(lib)
+        # FIXME: We should be able to inline this, that we can't means
+        #        we are inserting calls to it late.
+        LLVM.name(f) == "__ockl_hsa_signal_store" && continue
+        LLVM.name(f) == "__ockl_hsa_signal_load" && continue
+        attrs = function_attributes(f)
+        push!(attrs, EnumAttribute("alwaysinline"; ctx))
+    end
```
Comment (jpsamaroo) [inline src/device/runtime.jl:94]: Is it always valid and more performant to inline everything?

**Review by jpsamaroo — APPROVED** [review]

**Comment by jpsamaroo** [comment]
Any ideas why CI hung without any sort of async queue error?

File: src/device/runtime.jl (hunk 1 above)
Comment (vchuravy) [reply src/device/runtime.jl:94]: Should be yes. The functions that are exempt are used late and inlining removes them from the module.

File: src/device/gcn/math.jl [hunk 2]
```diff
@@ -14,8 +12,19 @@ for jltype in (
...
-    ); inp_args=(jltype,), out_arg=jltype))
+        :tgamma, :j0, :j1, :y0, :y1)
+
+        intrinsic == :expm1 && jltype == Float16 && continue
+        intrinsic == :erfinv && jltype == Float16 && continue
+        intrinsic == :erfcinv && jltype == Float16 && continue
+
+        if intrinsic == :sin && jltype == Float16
+            continue
+            push!(MATH_INTRINSICS, GCNIntrinsic(intrinsic, inp_args=(jltype,), out_arg=jltype, isbroken=true))
```
[the remark this answers is not in the document]
Comment (vchuravy) [reply src/device/gcn/math.jl:23]: This should work. So I rather have it here as a reminder that we need to figure why it doesn't

File: src/compat.jl [hunk 3]
```diff
@@ -64,3 +61,22 @@ function Statistics.corzm(x::ROCArray{<:Any, 2}, vardim::Int=1)
...
+import Random
+function Random.randn!(rng::GPUArrays.RNG, A::ROCArray{T}) where T<:Union{Float16, Float32, Float64}
+    threads = (length(A) - 1) ÷ 2 + 1
+    length(A) == 0 && return
+    gpu_call(A, rng.state; elements = threads) do ctx, a, randstates
+        idx = 2*(GPUArrays.linear_index(ctx) - 1) + 1
+        U1 = GPUArrays.gpu_rand(T, ctx, randstates)
+        U2 = GPUArrays.gpu_rand(T, ctx, randstates)
+        Z0 = sqrt(T(-2.0)*log(U1))*cos(T(2pi)*U2)
+        Z1 = sqrt(T(-2.0)*log(U1))*sin(T(2pi)*U2)
```
[the remark this answers is not in the document]
Comment (vchuravy) [reply src/compat.jl:75]: The one stolen from GPUArrays proper xD

**Comment by vchuravy** [comment]
> Any ideas why CI hung without any sort of async queue error?

I think we might simply be running up against the 1h time-limit?

**Comment by jpsamaroo** [comment]
That shouldn't be the case, the tests usually only take about 30 mins

File: src/device/gcn/math.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpsamaroo) [reply src/device/gcn/math.jl:23]: Sounds good! Will leave this unresolved for now, or if you want to add a FIXME comment and resolve this, that would be fine too.

File: src/compat.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpsamaroo) [reply src/compat.jl:75]: Ahh gotcha; might be good to just link to where it's taken from in a comment?

**Comment by vchuravy** [comment]
> That shouldn't be the case, the tests usually only take about 30 mins

```
Test Summary: | Pass  Broken  Total
AMDGPU        | 5597     115   5712
     Testing AMDGPU tests passed 

________________________________________________________
Executed in   54.66 mins    fish           external
   usr time   51.03 mins  373.00 micros   51.03 mins
   sys time    3.62 mins   34.00 micros    3.62 mins
```

And that is locally.

File: .buildkite/pipeline.yml [hunk 4]
```diff
@@ -13,7 +13,7 @@ steps:
       queue: "juliagpu"
       rocm: "*"
     if: build.message !~ /\[skip tests\]/
-    timeout_in_minutes: 60
+    timeout_in_minutes: 90
```
Comment (vchuravy) [inline .buildkite/pipeline.yml:16]:
```suggestion
    timeout_in_minutes: 120
```

File: .buildkite/pipeline.yml [hunk 5]
```diff
@@ -36,7 +36,7 @@ steps:
       queue: "juliagpu"
       rocm: "*"
     if: build.message !~ /\[skip docs\]/
-    timeout_in_minutes: 60
+    timeout_in_minutes: 90
```
Comment (vchuravy) [inline .buildkite/pipeline.yml:39]:
```suggestion
    timeout_in_minutes: 120
```

**Pull request merged**